### DISCUSSION TITLE
Use tonic server inside a tokio select loop?

### DISCUSSION THREAD

**Discussion by lquerel** [hyperium/tonic#934] [Q&A]
**Context** 
I have an application that exposes different types of APIs (REST, gRPC, ...). Each exposed API runs in a receiver task. A mpsc receiver is provided to each receiver to receive commands from an application controller. For example the app controller can use the mpsc channels to send a stop command to every receivers to implement graceful shutdown. 
I'd like to combine the mpsc receiver with a tonic server inside a tokio select loop in order to interleave commands and grpc calls (see pseudo code below). I didn't find an equivalent of my pseudo method `accept` without recreating entirely the tonic server in every iteration. 

Pseudo code (not working)
```rust
loop {
  let recv_future = receiver.recv();
  let server_future = tonic_server.accept();

  pin_mut!(recv_future, server_future);

  match select(recv_future, server_future).await {
    Either::Left((cmd, _)) => { /* do something with the command */ }
    Either::Right((grpc_call, _)) => { /* do something with the grpc call */ }
  }
}
``` 

**Questions**
* How can I achieve this kind of interleaving of 2 futures inside a tokio select loop? 
* If not possible, is there another way to achieve this?
* How can I implement a graceful shutdown of a tonic server part of a bigger app?

**Comment by LucioFranco** [comment]
I would suggest looking at the hyper example and only spawning a single hyper server. Hyper's builders expose a `graceful_shutdown` fn that you can pass a future to to start the process. I don't quite follow in your example code where `accept` comes from and you shouldn't need to use select on the server side to achieve what you want.

**Reply by lquerel** [reply]
Thanks - I finally discovered the method 'serve_with_incoming_shutdown' in Server::Builder() which fits my need well.
